offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy

```diff
@@ -0,0 +1,370 @@
+        for (OfferItem offerItem in offer.getDataAnalysisItems()) {
+            OfferPosition offerPosition = OfferPosition.createProductItem(position, offerItem)
+            tableBody.append(offerPosition.outerHtml())
+            position++
+            if (offerItem.getQuantityDiscount() > 0) {
+                OfferPosition discountPosition = OfferPosition.createDiscount(position, position - 1, offerItem)
+                tableBody.append(discountPosition.outerHtml())
+                position++
+            }
+        }
+        // set the footer
+        String netCosts = Currency.format(offer.getNetDataAnalysis())
+        dataAnalysisTable.select(" > tfoot .costs").first().text(netCosts)
+    }
+    private static void fillDataManagementItems(Document document, OfferContent offer) {
+        Element dataManagementTable = document.getElementById("data-management-items")
+        // remove table if no items are available
+        if (offer.getDataManagementItems().isEmpty()) {
+            dataManagementTable.remove()
+            return
+        }
+        // clear the table
+        dataManagementTable.select("tbody > tr").each {element ->
+            element.remove()
+        }
+        // fill the table
+        int position = 1
+        Element tableBody = dataManagementTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataManagementItems()) {
+            OfferPosition offerPosition = OfferPosition.createProductItem(position, offerItem)
+            tableBody.append(offerPosition.outerHtml())
+            position++
+            if (offerItem.getQuantityDiscount() > 0) {
+                OfferPosition discountPosition = OfferPosition.createDiscount(position, position - 1, offerItem)
+                tableBody.append(discountPosition.outerHtml())
+                position++
+            }
+        }
+        // set the footer
+        String netCosts = Currency.format(offer.getNetPMandDS())
+        document.select("#data-management-items > tfoot .costs").first().text(netCosts)
+    }
+
+    private static class OfferPosition {
+        private final int position
+        private final String name
+        private final String description
+        private final double quantity
+        private final double unitPrice
+        private final double total
+        private final String unit
+
+        static OfferPosition createProductItem(int position, OfferItem offerItem) {
+            String name = offerItem.getProductName()
+            String description = createProductItemDescription(offerItem.getProductDescription(), offerItem.getServiceProvider())
+            double quantity = offerItem.getQuantity()
+            double unitPrice = offerItem.getUnitPrice()
+            double total = offerItem.getItemTotal()
+            String unit = offerItem.getUnit()
+            return new OfferPosition(position, name, description, quantity, unitPrice, total, unit)
+        }
```

Comment:
see above, I dont think we need that

Reply:
I think we do. We would have to move the `outerHtml` code here if we did not have the instance with the fields set.
The `createProductItem` or `createDiscountItem` methods would have a String return value. This would not allow us to decouple the creation and the printing of the information.

Reply:
But we decided to not distinguish between offer item and discount. I see now what you have intended to do, but I think this is not necessary. We create the HTML code here and just want to set the fields. I would also argue we "created" the information for the HTML already when we created the offer content.